### ISSUE TITLE
Quantization Aware Training model has weird inference behavior

### ISSUE THREAD

**Issue by israelwei** [tensorflow/tensorflow#48212]
Hi,
I have a pretrained detection model I trained in Tensorflow 2.3 with fp32 precision. I used this model's weights as initial pretrained weights for Quantization Aware Training (QAT). During training I could see that training converged and gave logical predictions (I visualized results during training).
When trying to load with Tensorflow the Quantization Aware Training weights for inference the model behaves differently: When applying in inference the same preprocessing + normalization of mean/std, the predictions' probabilities are always around 0.25. Also, all the predictions' bounding boxes in inference are really small: around 15 pixels height/width, while during training I could see predictions having logical bounding boxes that are with varied dimensions and with much higher probabilities. This behavior appears even on images that were used for training so it's not related to overfitting issues.

Technical constraints for my implementation:
1) I used Tensorflow's `tf.quantization.quantize_and_dequantize` in specific places in my model: After the activation function which appears in the following pattern: `Conv2D` --> BatchNormalization --> `Activation`. I attach here an example for such pattern: 

```
from tensorflow.keras.layers import Conv2D, ZeroPadding2D, LeakyReLU, ReLU, BatchNormalization

def MyConv(x, filters, kernel_size, strides=1, batch_norm=True, is_quantized=False):
    if strides == 1:
        padding = "same"
    else:
        x = ZeroPadding2D(((1, 0), (1, 0)))(x)  # top left half-padding
        padding = "valid"
    x = Conv2D(filters=filters, kernel_size=kernel_size,
               strides=strides, padding=padding,
               use_bias=not batch_norm, kernel_regularizer=l2(0.0005))(x)

    if batch_norm:
        x = BatchNormalization()(x)
        if is_quantized:
            # We'll use The leaky version of Relu6 for stability of low-precision computations
            # We define the Leaky Relu using Relu for conversion constrains to TensorRT later
            x = ReLU(negative_slope=0.1, max_value=6)(x)
        else:
            x = LeakyReLU(alpha=0.1)(x)
    if is_quantized:
        # Using tf syntax of quantization - applying quantization only on the conv op, after the activation
        x = tf.quantization.quantize_and_dequantize(x, input_min=-64, input_max=64, range_given=False)
        # The input_min / input_max should be ignored due to range_given=False
    return x
```
This function is called when building the model both for training and inference. 
The reason I place the `quantize_and_dequantize` op specifically after the activation function is because this model should be converter to TensorRT later- their [QAT guidelines](https://docs.nvidia.com/deeplearning/frameworks/tf-trt-user-guide/index.html#quantization-training) note that this is the place to apply quantization op for QAT, due to the layer fusion TensorRT applies in the conversion and optimization process.
I tried also applying QAT using the [Tensorflow Model Optimization Toolkit as describes here](https://www.tensorflow.org/model_optimization/guide/quantization/training_example), but this resulted with a model that includes unsupported nodes for TensorRT such as `ConvInteger`. A TF model that uses quantize_and_dequantize produces model that is possible to convert to TRT, but has the issues I described above due to the irregular behavior in TF during inference. Here is an example to the quant ops in Netron for the inference model:

![qat structure example](https://user-images.githubusercontent.com/23454156/113120276-17bd0280-921a-11eb-9d79-9b0a80e8e7b9.PNG)

This is a Tensorflow-related issue since the issue appears during inference in Tensorflow with QAT weights.
2) I create the model for inference using the following command:
```
tf.keras.backend.set_learning_phase(1)
self.model = self._build_model()
```
where `build_model()` calls internally also the function I attached here which contains the quantization ops after every Activation function.

Is there any additional command that needs to be applied in inference when using `tf.quantize.quantize_and_dequantize` ? I remember in TF 1.x there was `tf.contrib.create_eval_graph`, but I didn't see anything similar to it in TF 2.3.
I prefer a solution that lets me still use `tf.quantize.quantize_and_dequantize` due to the TensorRT constraints I described (since I can't use Model Optimization Toolkit's `quantize_model`).

**System information**
- OS Platform and Distribution: Linux Ubuntu 18.04
- TensorFlow version: 2.3.0
- Python version: 3.6

**Comment by abattery** [comment]
@xhae  could you triage this issue?

**Comment by xhae** [comment]
@Xhark Can you take the issue?

**Comment by israelwei** [comment]
Another note, I would like to know if this is a possible cause for issues: When I inspect the onnx converted model from my QAT tensorflow model, I can see the following for all `QuantizeLinear` and `DequantizeLinear` nodes: `x_scale` values are always ~0.5039 (`x_zero_point` is 0, as it should be). I would have expected this quantization scale to be different for all layers since this is a learnable parameter for quantization op. Is it ok?

**Comment by israelwei** [comment]
I think I've found the issue- it's related to the way I loaded the weights with the relevant flags.
Previously I loaded the weights like this:
```
keras_model.load_weights(transform_weight_file, by_name=True)
```
Apparently the weights weren't loaded properly, and no warning or error was raised.
I changed it to:
```
keras_model.load_weights(transform_weight_file, by_name=False, skip_mismatch=False)
```
And now the inference values make sense for the QAT model both in probabilities and bounding boxes dimensions.

**Comment by Xhark** [comment]
Hi Would you please give me more context about how did you apply QAT? (code, how much step you train, etc...)
`x_scale` should changed during the training. (but depends on dataset)

Hi @daverim, do you have any context about TensorRT supports?

**Comment by israelwei** [comment]
It seems this is happening due to a tf2onnx issue- the `range_given=False` argument is currently ignored:
https://github.com/onnx/tensorflow-onnx/issues/1438#issuecomment-814516723
So the settings in Tensorflow were OK.

**Comment by daverim** [comment]
Seems that this is the issue -- looks like the scale is calculated from `input_min = -64, input_max = 64` (64/127 = .503937). Curious if the rest of this workflow is working correctly?

**Comment by israelwei** [comment]
The rest of the workflow is working correctly :) Only the scale was calculated in an wanted way

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
